sources/host-ctr/cmd/host-ctr/main.go

```diff
@@ -664,19 +669,68 @@ func tagImage(ctx context.Context, imageName string, newImageName string, client
 
 // withDynamicResolver provides an initialized resolver for use with ref.
 func withDynamicResolver(ctx context.Context, ref string) containerd.RemoteOpt {
-	if !strings.HasPrefix(ref, "ecr.aws/") {
-		// not handled here
-		return func(_ *containerd.Client, _ *containerd.RemoteContext) error { return nil }
-	}
-
-	return func(_ *containerd.Client, c *containerd.RemoteContext) error {
-		// Create the ECR resolver
-		resolver, err := ecr.NewResolver()
+	noOpt := func(_ *containerd.Client, _ *containerd.RemoteContext) error { return nil }
+
+	switch {
+	// For ECR registries, we need to use the Amazon ECR resolver
+	case strings.HasPrefix(ref, "ecr.aws/"):
+		return func(_ *containerd.Client, c *containerd.RemoteContext) error {
+			// Create the Amazon ECR resolver
+			resolver, err := ecr.NewResolver()
+			if err != nil {
+				return errors.Wrap(err, "Failed to create ECR resolver")
+			}
+			log.G(ctx).WithField("ref", ref).Info("pulling with Amazon ECR Resolver")
+			c.Resolver = resolver
+			return nil
+		}
+	// For Amazon ECR Public registries, we should try and fetch credentials before resolving the image reference
+	case strings.HasPrefix(ref, "public.ecr.aws/"):
+		// Try to get credentials for authenticated pulls from ECR Public
+		session := session.Must(session.NewSession())
+		// The ECR Public API is only available in us-east-1 today
+		publicConfig := aws.NewConfig().WithRegion("us-east-1")
+		client := ecrpublic.New(session, publicConfig)
+		output, err := client.GetAuthorizationToken(&ecrpublic.GetAuthorizationTokenInput{})
 		if err != nil {
-			return errors.Wrap(err, "Failed to create ECR resolver")
+			log.G(ctx).Warn("ecr-public: failed to get authorization token, falling back to default resolver (unauthenticated pull)")
+			return noOpt
 		}
-		log.G(ctx).WithField("ref", ref).Info("pulling with Amazon ECR Resolver")
-		c.Resolver = resolver
-		return nil
+		if output == nil || output.AuthorizationData == nil {
+			log.G(ctx).Warn("ecr-public: missing AuthorizationData in ECR Public GetAuthorizationToken response, falling back to default resolver (unauthenticated pull)")
+			return noOpt
+		}
+		authToken, err := base64.StdEncoding.DecodeString(aws.StringValue(output.AuthorizationData.AuthorizationToken))
+		if err != nil {
+			log.G(ctx).Warn("ecr-public: got invalid authorization token, falling back to default resolver (unauthenticated pull)")
+			return noOpt
+		}
+		tokens := strings.SplitN(string(authToken), ":", 2)
+		if len(tokens) != 2 {
+			log.G(ctx).Warn("ecr-public: got invalid authorization token, expecting token of the form 'username:password', falling back to default resolver (unauthenticated pull)")
+			return noOpt
+		}
+		// Use the fetched authorization credentials to resolve the image
+		authOpt := docker.WithAuthCreds(func(host string) (string, string, error) {
+			// Double-check to make sure the we're doing this for an ECR Public registry
+			if !strings.HasPrefix(host, "public.ecr.aws") {
```

Comment:
Do you just get the hostname as a parameter here or is it the full image?

If it's just the hostname, this would be better:
```suggestion
			if host != "public.ecr.aws" {
```

If it's the full image name, this would be better:
```suggestion
			if !strings.HasPrefix(host, "public.ecr.aws/") {
```

Reply:
It's just the hostname. I'll go ahead and fix it.